DM-21451: Remove DatabaseDict and vectorize Datastore/Butler ingest

TallJimbo opened the pull request.

TallJimbo committed:
Add external table functionality to Registry.

TallJimbo committed:
Use opaque tables instead of DatabaseDict in Datastores.

TallJimbo committed:
Remove DatabaseDict.

timj committed:
Ensure that each datastore in datastore tests has its own registry

Previously if a single test created two datastores they shared
a registry.

TallJimbo committed:
Rename DatasetExport to FileDataset.

We're about to start using it for ingest, so the old name no longer
made sense.

TallJimbo committed:
Vectorize GenericBaseDatastore info insertion.

TallJimbo committed:
Refactor and vectorize Datastore ingest.

TallJimbo committed:
Vectorize butler.ingest.

timj commented on python/lsst/daf/butler/core/registry.py:
```diff
@@ -197,32 +197,76 @@ def transaction(self):
+        """Add an opaque (to the `Registry`) table for use by a `Datastore` or
+        other data repository client.
+
+        Opaque table record can be added via `insertOpaqueData`, retreived via
```
Typo: retreived

timj commented on python/lsst/daf/butler/datastores/chainedDatastore.py:
```diff
@@ -305,97 +324,53 @@ def put(self, inMemoryDataset, ref):
-
-        # A "move" is sometimes a "copy"
-        moveIsCopy = False
-        if transfer == "move" and os.path.isabs(path):
```
You decided against keeping this logic?

TallJimbo replied:
Yeah.  At first I was just going to move it around, but it was a lot messier after the split into two methods, and it wasn't clear to me we had a real need for it.

timj commented on python/lsst/daf/butler/datastores/chainedDatastore.py:
```diff
@@ -305,97 +324,53 @@ def put(self, inMemoryDataset, ref):
-        notAcceptedCounter = 0
+        # Filter down to just datasets the chained datastore's own
+        # configuration accepts.
+        okForParent: List[FileDataset] = [dataset for dataset in datasets
```
If `okForParent` is empty shouldn't we raise DatasetTypeNotSupportedError ?

TallJimbo replied:
Raising that exception (in ingest) is now done exclusively by `Datastore.ingest`, which will check that the set of datasets actually ingested matches the set of datasets that the caller asked to be ingested and raise if that isn't true.  That's because we need `_prepIngest` to still return the list of datasets it *can* ingest even when it can't ingest all of them, and it can't do that if it raises.

timj replied:
I see that now. Thanks.

timj commented on python/lsst/daf/butler/datastores/chainedDatastore.py:
```diff
@@ -305,97 +324,53 @@ def put(self, inMemoryDataset, ref):
-            if moveIsCopy:
-                dstransfer = "copy"
+            if constraints is not None:
+                okForChild: List[FileDataset] = [dataset for dataset in okForParent
```
I'm a bit concerned that we've lost any visibility into whether any of the supplied datasets were ingested or not. Don't we worry if a dataset was rejected by all the child datastores because of constraints? As written it seems we worry about transfer modes but not constraints. If I ask 10 datasets to be ingested and only 9 are accepted how do I tell that? Previously ingest worked if a dataset was accepted by at least one datastore (and then we had the issue of deciding what to do if it was only accepted by an ephemeral datastore).

TallJimbo replied:
I think that's all still true (see reply at https://github.com/lsst/daf_butler/pull/197#discussion_r338674106).  Unlike the constraint-based rejections, transfer-mode rejections mean that all datasets would be rejected by the nested datastore, so we don't have to return anything and we can still use the raise-and-catch logic that was in place before.

timj commented on python/lsst/daf/butler/datastores/fileLikeDatastore.py:
```diff
@@ -202,51 +215,41 @@ def __init__(self, config, registry, butlerRoot=None):
+        # Docstring inherited from GenericBaseDatastore
+        records = list(self.registry.fetchOpaqueData(self._tableName, dataset_id=ref.id))
+        if len(records) == 0:
+            raise KeyError("Unable to retrieve formatter associated with Dataset {}".format(ref.id))
```
I don't think that error message should be referring to a formatter.

TallJimbo replied:
I've just moved it here from https://github.com/lsst/daf_butler/pull/197/commits/797a3dfe93f6a1435f2a0ec5201fc1b6535271ab#diff-4b3b7a0376839e94233cd3a2bddd157bL112, but I think I agree.  `"Unable to retrieve location associated with Dataset {}"` instead?

timj commented on python/lsst/daf/butler/datastores/fileLikeDatastore.py:
```diff
@@ -365,27 +368,107 @@ def _prepare_for_put(self, inMemoryDataset, ref):
+        # Docstring inherited from Datastore._prepIngest.
+        filtered = []
+        for dataset in datasets:
+            if not self.constraints.isAcceptable(dataset.ref):
```
Again how do we know that some of the datasets were not ingested?

TallJimbo replied:
(I think this is already answered above)

timj commented on python/lsst/daf/butler/datastores/genericDatastore.py:
```diff
@@ -121,29 +82,33 @@ def removeStoredItemInfo(self, ref):
-        #       self.registry internally.  Probably need to add
-        #       transactions to DatabaseDict to do better than that.
-        self.addStoredItemInfo(ref, itemInfo)
+        for ref, itemInfo in zip(refs, itemInfos):
```
I don't think zip is going to complain if the number of refs is not equal to the number of infos. Would it be safer to accept a sequence of tuples instead?

timj reviewed (approved):
This looks good in general. My main concern is that I'm not entirely clear we have a story for cases where some files are not ingested due to constraints. Previously this triggered DatasetTypeNotSupporterdError but I can't see where that happens now (maybe it does and I've missed it).

TallJimbo committed:
Switch _register_datasets sig. to sequence-of-tuples.

This is safer than zipping over two sequences, which doesn't complain
if they have different lengths, and no harder for callers to provide.

TallJimbo committed:
Improve error message for StoredFileInfo lookup failure.

(and switch to f-strings)
